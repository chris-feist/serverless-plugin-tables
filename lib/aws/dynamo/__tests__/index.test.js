const Plugin = require('../../..');

const dynamo = require('..');

beforeEach(() => {
  jest.clearAllMocks();
});

const dynamoService = {
  provider: {
    name: 'aws',
  },

  custom: {
    tables: {
      dynamo: {
        deploymentBatchSize: 2,
      },
    },
  },

  resources: {
    tables: {
      Music: {
        partitionKey: 'Artist',
        sortKey: 'SongTitle',
        indexes: [
          {
            name: 'GenreByTitleIndex',
            partitionKey: 'Genre',
            sortKey: 'AlbumTitle',
          },
        ],
      },
      People: {
        resourceName: 'FavoritePeopleDynamoTable',
        partitionKey: 'personID',
        sortKey: 'state',
        readUnits: 5,
        writeUnits: 5,
        indexes: [
          {
            name: 'EmailIndex',
            partitionKey: 'email',
            readUnits: 2,
            writeUnits: 2,
            projection: 'all',
          },
          {
            name: 'PersonByCreatedTimeIndex',
            sortKey: {
              name: 'createdTime',
              type: 'number',
            },
            readUnits: 2,
            writeUnits: 2,
            projection: 'keys',
          },
          {
            name: 'PersonByAgeIndex',
            sortKey: {
              name: 'age',
              type: 'number',
            },
            readUnits: 2,
            writeUnits: 2,
            projection: [
              'dob',
              'firstName',
              'lastName',
            ],
          },
        ],
        streamType: 'newItem',
        ttlKey: 'expirationTime',
        encrypted: true,
        pointInTimeRecovery: true,
        tags: {
          STAGE: 'test',
          TEAM: 'backend',
        },
        template: {
          Properties: {
            ProvisionedThroughput: {
              ReadCapacityUnits: 1,
            },
          },
        },
      },
    },
    Resources: {
      ExistingDynamoDBTable: {
        Type: 'AWS::DynamoDB::Table',
        DependsOn: 'FavoritePeopleDynamoTable',
        Properties: {
          TableName: 'ExistingTable',
          BillingMode: 'PAY_PER_REQUEST',
          AttributeDefinitions: [
            {
              AttributeName: 'userID',
              AttributeType: 'S',
            },
            {
              AttributeName: 'email',
              AttributeType: 'S',
            },
            {
              AttributeName: 'createdTime',
              AttributeType: 'N',
            },
          ],
          KeySchema: [
            {
              AttributeName: 'userID',
              KeyType: 'HASH',
            },
            {
              AttributeName: 'email',
              KeyType: 'RANGE',
            },
          ],
          LocalSecondaryIndexes: [
            {
              IndexName: 'UsersTableEmailGSI',
              KeySchema: [
                {
                  AttributeName: 'userID',
                  KeyType: 'HASH',
                },
                {
                  AttributeName: 'createdTime',
                  KeyType: 'RANGE',
                },
              ],
              Projection: {
                ProjectionType: 'ALL',
              },
            },
          ],
        },
      },
    },
  },
};

describe('dynamo', () => {
  test('default export', () => {
    expect(dynamo).toBeDefined();
    expect(dynamo).toMatchSnapshot();
  });

  test('end to end', () => {
    const mockServerless = {
      getProvider: jest.fn(),
      service: dynamoService,
    };
    const mockOptions = {};

    const instance = new Plugin(mockServerless, mockOptions);

    instance.processTables();

    expect(mockServerless.service.resources).toMatchSnapshot();
  });
});
