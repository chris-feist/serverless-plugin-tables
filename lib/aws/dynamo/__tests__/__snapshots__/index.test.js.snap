// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dynamo default export 1`] = `
Object {
  "postProcessSteps": Array [
    [Function],
    [Function],
    [Function],
  ],
  "tableSteps": Array [
    [Function],
    [Function],
    [Function],
  ],
}
`;

exports[`dynamo end to end 1`] = `
Object {
  "Resources": Object {
    "ExistingDynamoDBTable": Object {
      "DependsOn": Array [
        "FavoritePeopleDynamoTable",
      ],
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "userID",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "email",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "createdTime",
            "AttributeType": "N",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "userID",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "email",
            "KeyType": "RANGE",
          },
        ],
        "LocalSecondaryIndexes": Array [
          Object {
            "IndexName": "UsersTableEmailGSI",
            "KeySchema": Array [
              Object {
                "AttributeName": "userID",
                "KeyType": "HASH",
              },
              Object {
                "AttributeName": "createdTime",
                "KeyType": "RANGE",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
          },
        ],
        "TableName": "ExistingTable",
      },
      "Type": "AWS::DynamoDB::Table",
    },
    "FavoritePeopleDynamoTable": Object {
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "personID",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "state",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "email",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "createdTime",
            "AttributeType": "N",
          },
          Object {
            "AttributeName": "age",
            "AttributeType": "N",
          },
        ],
        "BillingMode": "PROVISIONED",
        "GlobalSecondaryIndexes": Array [
          Object {
            "IndexName": "EmailIndex",
            "KeySchema": Array [
              Object {
                "AttributeName": "email",
                "KeyType": "HASH",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
            "ProvisionedThroughput": Object {
              "ReadCapacityUnits": 2,
              "WriteCapacityUnits": 2,
            },
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "personID",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "state",
            "KeyType": "RANGE",
          },
        ],
        "LocalSecondaryIndexes": Array [
          Object {
            "IndexName": "PersonByCreatedTimeIndex",
            "KeySchema": Array [
              Object {
                "AttributeName": "personID",
                "KeyType": "HASH",
              },
              Object {
                "AttributeName": "createdTime",
                "KeyType": "RANGE",
              },
            ],
            "Projection": Object {
              "ProjectionType": "KEYS_ONLY",
            },
            "ProvisionedThroughput": Object {
              "ReadCapacityUnits": 2,
              "WriteCapacityUnits": 2,
            },
          },
          Object {
            "IndexName": "PersonByAgeIndex",
            "KeySchema": Array [
              Object {
                "AttributeName": "personID",
                "KeyType": "HASH",
              },
              Object {
                "AttributeName": "age",
                "KeyType": "RANGE",
              },
            ],
            "Projection": Object {
              "NonKeyAttributes": Array [
                "dob",
                "firstName",
                "lastName",
              ],
              "ProjectionType": "INCLUDE",
            },
            "ProvisionedThroughput": Object {
              "ReadCapacityUnits": 2,
              "WriteCapacityUnits": 2,
            },
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": true,
        },
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 1,
          "WriteCapacityUnits": 5,
        },
        "SSESpecification": Object {
          "SSEEnabled": true,
        },
        "StreamSpecification": Object {
          "StreamViewType": "NEW_IMAGE",
        },
        "TableName": "People",
        "Tags": Array [
          Object {
            "Key": "STAGE",
            "Value": "test",
          },
          Object {
            "Key": "TEAM",
            "Value": "backend",
          },
        ],
        "TimeToLiveSpecification": Object {
          "AttributeName": "expirationTime",
          "Enabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
    },
    "MusicDynamoDbTable": Object {
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "Artist",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "SongTitle",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "Genre",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "AlbumTitle",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "GlobalSecondaryIndexes": Array [
          Object {
            "IndexName": "GenreByTitleIndex",
            "KeySchema": Array [
              Object {
                "AttributeName": "Genre",
                "KeyType": "HASH",
              },
              Object {
                "AttributeName": "AlbumTitle",
                "KeyType": "RANGE",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "Artist",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "SongTitle",
            "KeyType": "RANGE",
          },
        ],
        "TableName": "Music",
      },
      "Type": "AWS::DynamoDB::Table",
    },
  },
}
`;
